### ISSUE TITLE
LimeTabularExplainer gives an unexpected explanation

### ISSUE THREAD

**Issue by peterwvj** [marcotcr/lime#248]
Inspired by [this](https://marcotcr.github.io/lime/tutorials/Tutorial%20-%20continuous%20and%20categorical%20features.html) tutorial I've been trying to explain the results produced by a classification model that analyses network traffic. The model uses the "KDD Cup 99 data 10 %" data set, which is available [here](http://kdd.ics.uci.edu/databases/kddcup99/kddcup.data_10_percent.gz). It contains a mix of numerical and categorical features. Each row in the data set contains information about network connections. The last column contains a label that indicates whether the connection was *normal* or an *attack*. Examples of attack types include `smurf`(a kind of DoS attack), `buffer_overflow.` and many more. As an example, two samples are shown below (a smurf attack and a normal connection):

```
0,icmp,ecr_i,SF,1032,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,511,511,0.00,0.00,0.00,0.00,1.00,0.00,0.00,255,255,1.00,0.00,1.00,0.00,0.00,0.00,0.00,0.00,smurf.
0,tcp,http,SF,181,5450,0,0,0,0,0,1,0,0,0,0,0,0,0,0,0,0,8,8,0.00,0.00,0.00,0.00,1.00,0.00,0.00,9,9,1.00,0.00,0.11,0.00,0.00,0.00,0.00,0.00,normal.
```

 The problem is that I get unexpected explanations when I use the the `LimeTabularExplainer` to explain predictions. For example, when the model predicts that a given connection is a `smurf` attack, the explanation is based on some other attack (`buffer_overflow`). The following figure shows an example of this.

![explaination](https://user-images.githubusercontent.com/5388919/47351393-28b3d200-d6b8-11e8-8c04-48361e46a8ce.png)

In particular, I'm wondering why the explanation uses "NOT buffer_overflow" and "buffer_overflow" for the chart in the middle.

I was wondering whether this has sometimes to do with the order of the class names passed to the `LimeTabularExplainer`. In particular the documentation says "class_names – list of class names, ordered according to whatever the classifier is using. If not present, class names will be ‘0’, ‘1’, ..." but it is unclear to me how I find out what the order is.

Any help is appreciated. Thanks in advance!

The complete code is available below:

```python
import sklearn
import sklearn.datasets
import sklearn.ensemble
import numpy as np
import lime
import lime.lime_tabular

feature_names = [
    "duration",  # continuous.
    "protocol_type",  # symbolic.
    "service",  # symbolic.
    "flag",  # symbolic.
    "src_bytes",  # continuous.
    "dst_bytes",  # continuous.
    "land",  # symbolic.
    "wrong_fragment",  # continuous.
    "urgent",  # continuous.
    "hot",  # continuous.
    "num_failed_logins",  # continuous.
    "logged_in",  # symbolic.
    "num_compromised",  # continuous.
    "root_shell",  # continuous.
    "su_attempted",  # continuous.
    "num_root",  # continuous.
    "num_file_creations",  # continuous.
    "num_shells",  # continuous.
    "num_access_files",  # continuous.
    "num_outbound_cmds",  # continuous.
    "is_host_login",  # symbolic.
    "is_guest_login",  # symbolic.
    "count",  # continuous.
    "srv_count",  # continuous.
    "serror_rate",  # continuous.
    "srv_serror_rate",  # continuous.
    "rerror_rate",  # continuous.
    "srv_rerror_rate",  # continuous.
    "same_srv_rate",  # continuous.
    "diff_srv_rate",  # continuous.
    "srv_diff_host_rate",  # continuous.
    "dst_host_count",  # continuous.
    "dst_host_srv_count",  # continuous.
    "dst_host_same_srv_rate",  # continuous.
    "dst_host_diff_srv_rate",  # continuous.
    "dst_host_same_src_port_rate",  # continuous.
    "dst_host_srv_diff_host_rate",  # continuous.
    "dst_host_serror_rate",  # continuous.
    "dst_host_srv_serror_rate",  # continuous.
    "dst_host_rerror_rate",  # continuous.
    "dst_host_srv_rerror_rate"  # continuous.
]

data = np.genfromtxt(
    '/path/to/kddcup.data_10_percent_corrected',
    delimiter=',',
    dtype=str)
labels = data[:, -1]
le = sklearn.preprocessing.LabelEncoder()
le.fit(labels)
labels = le.transform(labels)
class_names = le.classes_
data = data[:, :-1]

categorical_features = [1, 2, 3, 6, 11, 20, 21]

categorical_names = {}
for feature in categorical_features:
    le = sklearn.preprocessing.LabelEncoder()
    le.fit(data[:, feature])
    data[:, feature] = le.transform(data[:, feature])
    categorical_names[feature] = le.classes_

data = data.astype(float)

encoder = sklearn.preprocessing.OneHotEncoder(
    categorical_features=categorical_features)

np.random.seed(1)
train, test, labels_train, labels_test = sklearn.model_selection.train_test_split(
    data, labels, train_size=0.80)

encoder.fit(data)
encoded_train = encoder.transform(train)

rf = sklearn.ensemble.RandomForestClassifier(n_estimators=10)
rf.fit(encoded_train, labels_train)

sklearn.metrics.accuracy_score(labels_test, rf.predict(encoder.transform(test)))

predict_fn = lambda x: rf.predict_proba(encoder.transform(x))

explainer = lime.lime_tabular.LimeTabularExplainer(
    train,
    feature_names=feature_names,
    class_names=class_names,
    categorical_features=categorical_features,
    categorical_names=categorical_names,
    kernel_width=3)

np.random.seed(1)
i = 1653
exp = explainer.explain_instance(test[i], predict_fn, num_features=5)
exp.save_to_file('/path/to/explanation.html')
```

**Comment by peterwvj** [comment]
Explicitly setting top_labels to 1 seem to solve the problem:

```python
exp = explainer.explain_instance(test[i], predict_fn, num_features=41, top_labels=1)
```

**Comment by marcotcr** [comment]
Sorry for the delay in responding. `explainer.explain_instance` has a parameter called `labels`, which specifies for which labels you want explanations. The default value is 1, which is what makes sense for binary classification. `top_labels` is an alternative that gets explanations for the top predicted labels, which is why that solves your problem. We should make top_labels be the default when the number of classes is > 2.

**Comment by peterwvj** [comment]
> We should make top_labels be the default when the number of classes is > 2.

Yes, I agree. Many thanks for explaining this.
